### PR TITLE
Consistent Hashing Support

### PR DESCRIPTION
**Pull request by rtkrruvinskiy**
Consistent hashing algorithm is same as in carbon.

Configuration syntax is
```
addRoute consistentHashing <route_name>  <server1>:<port1>:<instance1> <options...>  <server1>:<port1>:<instance1> <options...>  ...  <servern>:<portn>:<instancen> <options...>
```

Example:

addRoute consistentHashing cache  127.0.0.1:2013:a pickle=false  127.0.0.1:2023:b pickle=true

### CODE REVIEW

**Comment by rtkrruvinskiy** [comment]
I'm submitting this for initial review. I was going to also add tests to carbon-relay-ng_test.go, but some of the tests there fail for me even without my changes (e.g., Test2EndpointsUp fails with "FATAL: tE :2005 timed out after 50ms of waiting for 1 accepts (saw 0)FATAL: tE :2006 timed out after 50ms of waiting for 1 accepts (saw 0)"). If you're good with my general approach, I can try to track down the existing failures I'm seeing and put in additional tests.

**Comment by Dieterbe** [comment]
100% compatible with carbon?? very nice, i know some people were asking about that.

initial review looks allright, didn't look in depth yet. 
yeah those existing tests can be problematic sometimes, i also ran into that occasionally, i never truly figured out what was up, usually increasing the timeouts worked, it seemed to also depend on other stuff going on your computer.

**Comment by rtkrruvinskiy** [comment]
It should be 100% compatible with Carbon, as I translated the Python implementation. The unit test for the hasher tries to ensure that. It is possible there's an edge case I've overlooked, but if so, it should be easy to fix.

File: destination.go [hunk 1]
```diff
@@ -13,12 +13,24 @@ func addrToPath(s string) string {
...
 }
 
+func addrInstanceSplit(addr string) (string, string) {
+	var instance string
+	// The address may be specified as server:port or server:port:instance.
+	addrComponents := strings.Split(addr, ":")
+	if len(addrComponents) == 3 {
+		instance = addrComponents[2]
+	}
+	addr = strings.Join(addrComponents[0:2], ":")
```
Comment (Dieterbe) [inline destination.go]: could this not result in a slice out of bounds error? if original input had no ':'

File: route.go [hunk 2]
```diff
@@ -147,53 +188,108 @@ func (route *baseRoute) Shutdown() error {
...
+// on a configuration change.
+// The baseRoute object implements private methods (addDestination,
+// delDestination, etc.) that create a new baseRouteConfig object that reflects
+// the configuration change. This baseRouteConfig object is applicable to all
+// route types. However, some route types, like ConsistentHashingRoute, have
+// additional configuration and therefore have a distinct configuration object
+// that embeds baseRouteConfig (in the case of ConsistentHashingRoute, the object
+// is consistentHashingRouteConfig). This route-type-specific configuration also
+// needs to be updated on a base configuration change (e.g., on a change
+// affecting destinations). Accordingly, the public entry points that effect the
```
Comment (Dieterbe) [inline route.go]: affect

**Comment by Dieterbe** [comment]
please update the readme. just look for "consistent hashing" in it and you'll find where it needs updating ;) it should definitely be listed under features, and also mention explicitly it's designed to be 100% compatible with carbon's consistent hashing logic.

File: route.go [hunk 3]
```diff
@@ -6,9 +6,27 @@ import (
...
+func (c baseRouteConfig) Matcher() *Matcher {
+	return &c.matcher
+}
+
+func (c baseRouteConfig) Dests() []*Destination {
+	return c.dests
+}
+
+type consistentHashingRouteConfig struct {
+	RouteConfig
```
Comment (Dieterbe) [inline route.go]: i didn't know go lets you embed an interface type? i thought you'd embed baseRouteConfig here. how does this work?

File: destination.go (hunk 1 above)
Comment (rtkrruvinskiy) [inline destination.go]: You're right! I forgot about the case where a port is not specified. What port ends up being used by default, btw?

File: route.go (hunk 3 above)
Comment (rtkrruvinskiy) [inline route.go]: Er, yes, I did mean to embed baseRouteConfig. Apparently, though, [embedding interfaces is allowed](http://stackoverflow.com/questions/24537443/meaning-of-a-struct-with-embedded-anonymous-interface), and it just means I can initialize the field with anything implementing the interface. Nonetheless, I'll change this to baseRouteConfig to avoid confusion (and to do what I had originally intended).

File: route.go (hunk 2 above)
Comment (rtkrruvinskiy) [inline route.go]: I really did mean effect as a verb in this case. :) http://dictionary.reference.com/browse/effect, meaning 10.

File: route.go [hunk 4]
```diff
@@ -103,19 +137,26 @@ func (route *RouteSendFirstMatch) Dispatch(buf []byte) {
 	}
 }
 
+func (route *RouteConsistentHashing) Dispatch(buf []byte) {
+	conf := route.config.Load().(consistentHashingRouteConfig)
+	dest := conf.Dests()[conf.Hasher.GetDestinationIndex(buf)]
+	log.Info("route %s sending to dest %s: %s", route.key, dest.Addr, buf)
```
Comment (Dieterbe) [inline route.go]: needs a `// dest should handle this as quickly as it can` comment like the other routes, just to be clear.

File: route.go [hunk 5]
```diff
@@ -147,53 +188,108 @@ func (route *baseRoute) Shutdown() error {
...
+// needs to be updated on a base configuration change (e.g., on a change
+// affecting destinations). Accordingly, the public entry points that effect the
+// configuration change (Add, DelDestination, etc.) are implemented for baseRoute
+// and also for any route type, like ConsistentHashingRoute, that creates a
+// configuration object. These public entry points call the private method,
+// passing in a callback function of type baseConfigExtender, which takes a
+// baseRouteConfig and either returns it unchanged or creates an outer
+// configuration object with the baseRouteConfig embedded in it.
+// The private method then stores the RouteConfig object returned by the callback.
+type baseConfigExtender func(baseRouteConfig) RouteConfig
```
Comment (Dieterbe) [inline route.go]: interesting approach.  i was thinking maybe this a bit too much indirection/complication, i think i would just have the internal methods return the baseRouteConfig and have all the public methods (for example `RouteConsistentHashing.DelDestination`) take that return value and do any "config post processing", and then store it. this would admittedly have a bit more repetitive code but it might be easier to understand.

either way, i'm not opposed to your approach.  if you like it, it's fine with me.

File: route.go (hunk 2 above)
Comment (Dieterbe) [inline route.go]: oh cool. learned something new today :)

File: destination.go (hunk 1 above)
Comment (Dieterbe) [inline destination.go]: you know, i don't know. maybe the addr is already pre-validated somewhere, i don't remember. can you figure it out? :p

File: route.go (hunk 3 above)
Comment (Dieterbe) [inline route.go]: wow! i never knew you could do this. "partially overriding" an interface implementation is something i thought was impossible, but look at this.. so elegant.

File: route.go (hunk 5 above)
Comment (rtkrruvinskiy) [inline route.go]: Yeah, I can sometimes probably be excessively preoccupied with avoiding code repetition.

I agree that this approach is a little unusual, but if you're okay with it being this way, I'd rather keep it.

**Comment by rtkrruvinskiy** [comment]
I've updated the README, made the changes we talked about, and also changed the hasher and matcher to be stored by pointer (rationale described in the commit message).

Additionally, I've fiddled with the timing in the existing tests to get them to pass for me reliably and have added routing tests for consistent hashing.

**Comment by Dieterbe** [comment]
Great work!

> Store Matcher by pointer in baseRouteConfig. It is used by pointer almost
  anywhere in practice, and this allows us to get rid of several dereferences.

with "used by pointer almost anywhere" do you mean the Matcher methods are pointer receivers?
maybe we should change those to value receivers. not sure. it might perform better than pointer lookups.
but the important thing to note here is that currently, the Matcher properties of all other types (destination, route, routesnapshots) are values, not pointers. and i really think we should not introduce inconsistencies like this, it can be quite confusing.

another bit of history that I think is perhaps not super related, but useful nonetheless:
it used to be that matchers were mutable properties (like everything else) and so had update methods.
than at one point @pwielgolaski noted it would make sense to make them immutable (https://github.com/graphite-ng/carbon-relay-ng/issues/22#issuecomment-59618077) and which was committed in https://github.com/graphite-ng/carbon-relay-ng/commit/71870c5294875bc21111c80bb04415308a59d59f
since that change, we just create and assign new matcher objects when we need different matching logic.

**Comment by rtkrruvinskiy** [comment]
I ran the following benchmark code:
```go
func BenchmarkLoadConfig(b *testing.B) {
	cmd := "addRoute sendAllMatch test  127.0.0.1:2005  127.0.0.1:2006  127.0.0.1:2007"
	table := NewTableOrFatal(nil, "", cmd)
	value := table.config.Load().(TableConfig).routes[0].(*RouteSendAllMatch).config
	for i := 0; i < b.N; i++ {
		_ = value.Load().(RouteConfig).Matcher()
	}
}
```
When Matcher is stored by value, the results are:
BenchmarkLoadConfig	20000000	        92.4 ns/op
When Matcher is stored by pointer, the results are:
BenchmarkLoadConfig	50000000	        34.2 ns/op

What would you prefer that I do? One thing that may make sense is to undo that part of the commit here since it really has nothing to do with Consistent Hashing and then take it up under a different issue.

**Comment by Dieterbe** [comment]
> One thing that may make sense is to undo that part of the commit here since it really has nothing to do with Consistent Hashing and then take it up under a different issue.

this sounds best to me.  introduce CH in a way that blends in nicely with the code, and address Matcher pointers in a separate issue.

**Comment by rtkrruvinskiy** [comment]
Done.

**Comment by Dieterbe** [comment]
great work! happy to see someone tackle this.

**Comment by jmvitali** [comment]
hello,
Could you provide an example of consistentHashing route with replicaCount option?
Thanks

**Comment by rtkrruvinskiy** [comment]
@jmvitali Sorry, but my PR didn't include support for replica counts. The behaviour of the code is equivalent to a replica count of 1.

**Comment by jmvitali** [comment]
@rtkrruvinskiy thanks for the quick answer.
